base/set.jl

```diff
@@ -191,6 +191,65 @@ function unique(f::Callable, C)
 end
 
 """
+    unique!(A::AbstractVector)
+
+Remove duplicate items as determined by [`isequal`](@ref), then return the modified `A`.
+
+```jldoctest
+julia> unique!([1, 1, 1])
+1-element Array{Int64,1}:
+ 1
+
+julia> A = [7, 3, 2, 3, 7, 5];
+
+julia> unique!(A);
+
+julia> A
+4-element Array{Int64,1}:
+ 7
+ 3
+ 2
+ 5
+```
+"""
+function unique!(A::AbstractVector)
+    if isempty(A)
+        return A
+    elseif issorted(A) || issorted(A, rev=true)
+        # If A is sorted, then we only need to keep track of one element and add that to A
+        # every time that we see a new element.
+        idxs = eachindex(A)
+        m = start(idxs)
+        for i in idxs
+            x = A[i]
+            if x != A[m]
+                _, m = next(idxs, m)
```

Comment:
In general doing this isn't correct: `m` is the state of the iterator, which can be different from the index. So you need to store both the index and state in separate variables, as you do in the other branch. Hopefully in most cases the compiler will be able to elide redundant computations.